chainer/functions/connection/depthwise_convolution_2d.py

```diff
@@ -0,0 +1,164 @@
+import numpy
+from six import moves
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import conv
+from chainer.utils import type_check
+
+
+def _pair(x):
+    if hasattr(x, '__getitem__'):
+        return x
+    return x, x
+
+
+class DepthwiseConvolution2D(function.Function):
+
+    def __init__(self, stride=1, pad=0):
+        self.sy, self.sx = _pair(stride)
+        self.ph, self.pw = _pair(pad)
+
+    def check_type_forward(self, in_types):
+        n_in = in_types.size()
+        type_check.expect(2 <= n_in, n_in <= 3)
+
+        x_type = in_types[0]
+        w_type = in_types[1]
+        type_check.expect(
+            x_type.dtype.kind == 'f',
+            w_type.dtype.kind == 'f',
+            x_type.ndim == 4,
+            w_type.ndim == 4,
+            x_type.shape[1] == w_type.shape[1],
+        )
+
+        if n_in.eval() == 3:
+            b_type = in_types[2]
+            type_check.expect(
+                b_type.dtype == x_type.dtype,
+                b_type.ndim == 1,
+                b_type.shape[0] == w_type.shape[0] * w_type.shape[1],
+            )
+
+    def forward(self, inputs):
+        x, W = inputs[:2]
+        b = inputs[2] if len(inputs) == 3 else None
+        kh, kw = W.shape[2:]
+
+        xp = cuda.get_array_module(*x)
+        if xp is numpy:
+            self.col = conv.im2col_cpu(
+                x, kh, kw, self.sy, self.sx, self.ph, self.pw)
+        else:
+            self.col = conv.im2col_gpu(
+                x, kh, kw, self.sy, self.sx, self.ph, self.pw)
+
+        arys = [xp.tensordot(self.col[:, i, :, :, :, :], W[:, i, :, :],
+                             ((1, 2), (1, 2))).astype(x.dtype, copy=False)
+                for i in moves.range(W.shape[1])]
+
+        # along input channel axis
+        y = xp.concatenate(arys, axis=3)
+
+        if b is not None:
+            y += b
+        return xp.rollaxis(y, 3, 1),
+
+    def backward(self, inputs, grad_outputs):
+        x, W = inputs[:2]
+        b = inputs[2] if len(inputs) == 3 else None
+        gy = grad_outputs[0]
+        h, w = x.shape[2:]
+
+        xp = cuda.get_array_module(*x)
+        gy = xp.rollaxis(gy, 1, 4)
+        garys = xp.split(gy, W.shape[1], axis=3)
+        gW = xp.empty_like(W)
+        gcol = xp.empty_like(self.col)
+        gcol = xp.rollaxis(self.col, 0, 4)
+        for i in moves.range(W.shape[1]):
```

Comment:
Also we can get better performance when using xp functions instead of for-loop here.